chainerx_cc/chainerx/routines/manipulation.cc

```diff
@@ -781,19 +784,21 @@ Array Moveaxis(const Array& a, const Axes& source, const Axes& destination) {
     std::iota(source_axes.begin(), source_axes.end(), 0);
     std::iota(destination_axes.begin(), destination_axes.end(), 0);
 
-    for (uint8_t i = 0; i < source.ndim(); ++i) {
-        order[destination[i]] = source[i];
-        source_axes[source[i]] = -1;
-        destination_axes[destination[i]] = -1;
+    for (int8_t i = 0; i < source.ndim(); ++i) {
+        order[normalized_destination[i]] = normalized_source[i];
+        source_axes[normalized_source[i]] = -1;
+        destination_axes[normalized_destination[i]] = -1;
     }
 
-    source_axes.erase(std::remove(source_axes.begin(), source_axes.end(), -1), source_axes.end());
-    destination_axes.erase(std::remove(destination_axes.begin(), destination_axes.end(), -1), destination_axes.end());
-    CHAINERX_ASSERT(static_cast<int8_t>(source_axes.size()) == a.ndim() - source.ndim());
-    CHAINERX_ASSERT(static_cast<int8_t>(destination_axes.size()) == a.ndim() - destination.ndim());
+    Axes::iterator source_iter = std::remove(source_axes.begin(), source_axes.end(), -1);
+    Axes::iterator destination_iter = std::remove(destination_axes.begin(), destination_axes.end(), -1);
+    CHAINERX_ASSERT(static_cast<int8_t>(source_iter - source_axes.begin()) == a.ndim() - source.ndim());
+    CHAINERX_ASSERT(static_cast<int8_t>(destination_iter - destination_axes.begin()) == a.ndim() - destination.ndim());
 
-    for (int8_t i = 0; i < a.ndim() - source.ndim(); ++i) {
-        order[destination_axes[i]] = source_axes[i];
+    for (std::pair<Axes::iterator, Axes::iterator> i(source_axes.begin(), destination_axes.begin());
+         i.first != source_iter /* && i.second != destination_iter */;
+         ++i.first, ++i.second) {
+        order[*i.second] = *i.first;
```

Comment:
This for-loop seems complex. I think the previous version is better.

Reply:
Another solution is to use `std::transform`.
```cpp
std::transform(source_axes.begin(), source_iter, destination_axes.begin(), source_axes.begin(),
        [&](int8_t source_index, int8_t destination_index){
    return order[destination_index] = source_index;
});
```
However, I don't want to use it because `std::transform` requires memory spaces to set return the values. (The above implementation abuse `source_axes` also for output memory space.) Do you have any better ideas?

Reply:
Current is relatively complex, but it has minimum overhead I guess since we don't the size of any vector. We can add comments to describe the behaviour?

If we don't want to return as we need to in `transform`, we can use, `std::for_each` in a hacky way,
```CPP
int index = 0;
std::for_each(source_axes.begin(), source_iter, [&](auto& src_ele){ order[destination[index]] = src_ele; ++index; });
```

By previous do you mean using `vector.erase`?

Reply:
I mean the version of `order[destination_axes[i]] = source_axes[i]`. Does it take overhead? I think that the overhead from index accesses will be resolved by compiler optimization.

Reply:
We will need to do vector erase like earlier
```CPP
source_axes.erase(std::remove(source_axes.begin(), source_axes.end(), -1), source_axes.end());
destination_axes.erase(std::remove(destination_axes.begin(), destination_axes.end(), -1), destination_axes.end());
CHAINERX_ASSERT(static_cast<int8_t>(source_axes.size()) == a.ndim() - source.ndim());
CHAINERX_ASSERT(static_cast<int8_t>(destination_axes.size()) == a.ndim() - destination.ndim());

for (int8_t i = 0; i < a.ndim() - source.ndim(); ++i) {
        order[destination_axes[i]] = source_axes[i];
    }
```
 or if we don't want erase then have a separate variable for `index`,
```CPP
Axes::iterator source_iter = std::remove(source_axes.begin(), source_axes.end(), -1);
Axes::iterator destination_iter = std::remove(destination_axes.begin(), destination_axes.end(), -1);
CHAINERX_ASSERT(static_cast<int8_t>(source_iter - source_axes.begin()) == a.ndim() - source.ndim());
CHAINERX_ASSERT(static_cast<int8_t>(destination_iter - destination_axes.begin()) == a.ndim() - destination.ndim());

int idx = 0;
for (std::pair<Axes::iterator, Axes::iterator> i(source_axes.begin(), destination_axes.begin());
         i.first != source_iter /* && i.second != destination_iter */;
         ++i.first, ++i.second) {
        order[destination_axes[idx]] = source_axes[idx];
        idx++;
}
```
Or something else? :thinking:

Reply:
> We will need to do vector erase like earlier.
Why? I think there is no need to do because for-loop iterates only `a.ndim() - source.ndim()` times.

Reply:
You mean like this?

```CPP
Axes::iterator source_iter = std::remove(source_axes.begin(), source_axes.end(), -1);
Axes::iterator destination_iter = std::remove(destination_axes.begin(), destination_axes.end(), -1);

int8_t rest_dim = a.ndim() - source.ndim()
CHAINERX_ASSERT( a.ndim() - destination.ndim() == rest_dim);
CHAINERX_ASSERT(static_cast<int8_t>(source_iter - source_axes.begin()) == rest_dim);
CHAINERX_ASSERT(static_cast<int8_t>(destination_iter - destination_axes.begin()) == rest_dim);

for (int8_t i = 0; i < rest_dim; ++i) {
        order[destination_axes[i]] = source_axes[i];
    }
```

Reply:
Yes!

Reply:
Done.